Evaluate R code from JS with `evalRCode()`, and wrap R objects in proxies for the main thread

georgestagg opened the pull request:
This PR adds infrastructure for calling R code from JS and receiving the results as JS or RProxy objects, rather than working with strings directly from `stdin` and `stdout`.

An `evalRCode` function is patched into R that takes a string, evaluates it, and returns a `SEXP` pointer to the result. The `evalRCode` function optionally takes an environment as the second argument.

---

A new file `sexp.ts` contains a `RSexp` class, along with various subclasses for different `SEXP` types in R. These classes handle interpreting the `SEXP` pointers from R, allocating memory for new objects, reading R data from memory and converting to JS objects.

The `RSexp` constructor can take either a pointer to a `SEXP` object in WASM memory, reading and interpreting whatever is already there. Or, alternatively, take a JS object which is copied to WASM memory by allocating new R `SEXP` objects (where supported).

Some `RSexp` objects are setup to automatically convert their contents into a JS object when requested, deciding based on the `convertImplicitly` property on the subclass.

Many functions from R's `Rinternals.h` are used in this file, which are directly accessed through `Module`. It is intended that this work all happens on the worker thread.

Note that not everything is implemented yet, setters in particular. However, this feels like a good start.

---

A new file `proxy.ts` handles proxying of R `SEXP` objects. An `Rproxy` object on the main thread targets a reference to an R object in the worker thread, and these target references are given a type of `RTargetObj`. Requests for getting, setting, or applying properties to the targets are intercepted by the proxy traps and converted into messages to be sent to the worker thread.

`RTargetObj`s are designed to be simple enough to be passed around by `postMessage`. They are passed from the worker to the main thread ready to be the target of an`RProxy`, and then later these targets are re-interpreted as `RSexp` objects when they are passed back to the worker thread as part of a further request.

I have created three types of `RTargetObj`:
  - `RAW`: The object is a raw JS object that is able to be transferred via `postMessage`.
  - `SEXPPTR`: The object is a pointer to an R SEXP in WASM memory.
  - `CODE`: The object is the result of as yet unevaluated R code.

---

The worker now has new `getRObj` and `setRObj` request types, invoked by RProxy trapped object property requests. The requests take a `RTargetObj` passed from the main thread, create a `RSexp` object using that target, and then obtain the requested chain of properties from the `RSexp` object.

If a `CODE` target object is passed the associated R code is first evaluated, and then the requested property chain is retrieved. For `getRObj`, the final property in the chain can be optionally called if some call arguments are included in the request.

The result is either a raw JS object or reference to an R SEXP object in WASM memory, and is returned back to the main thread as a `RTargetObj` ready to be wrapped up in a new `RProxy`.

---

`channel.ts` has been tweaked slightly in how it handles responses to requests. If a 'response' type message comes in to the main thread the request promise is resolved immediately, rather than being added to the output queue. This avoids deadlocks where a response is being awaited on without a main thread inf-loop also clearing the output queue. 

---

Several files have been modified in a minor ways to keep `eslint` happy. Apologies for the noise.

---

Below are some fun examples of the things that can be done after this change, running in a JS console. I hope to include some real documentation of how to use the RProxy objects soon, in a future PR.

```
$ await webR.evalRCode('42')
> 42

$ await webR.evalRCode('c(1,2,3)')
> Float64Array(3) [1, 2, 3, buffer: ArrayBuffer(125435904), byteLength: 24, ...

$ await webR.evalRCode('c(TRUE, FALSE)')
> (2) [true, false]

$ await webR.evalRCode('c(TRUE, FALSE, "a")')
> (3) ['TRUE', 'FALSE', 'a']

$ await webR.evalRCode('NULL')
> undefined
$ await webR.evalRCode('NaN')
> NaN

$ await webR.evalRCode('list(a=17, b=18)').toJs()
> {a: 17, b: 18}

$ fn = await webR.evalRCode('function(x){x+1}')
> Proxy {obj: 62262528, type: 'SEXPPTR', prototype: undefined}
$ await fn(5)
> 6

$ s = await webR.evalRCode('sample')
> Proxy {obj: 62263456, type: 'SEXPPTR', prototype: undefined}
$ await s(['a','b','c'], 5, true)
> (5) ['b', 'a', 'b', 'a', 'b']

$ await webR.evalRCode('sin')(webR.evalRCode('c(1,2,3)'))
> Float64Array(3) [0.8414709848078965, 0.9092974268256817, 0.1411200080598672, ...

$ c = await webR.evalRCode('c(a=1,b=2)')
> Proxy {obj: 102415680, type: 'SEXPPTR', prototype: undefined}
$ c['b'] = 5
> 5
$ await c.toJs()
> {a: 1, b: 5}

$ env = webR.evalRCode('e=new.env(); e$x = 42; e')
> Proxy {obj: {…}, type: 'CODE', prototype: undefined}
$ await env.ls
> 'x'
$ await env.$x
> 42
$ await webR.evalRCode('x', env)
> 42
```

lionel- commented on src/webR/sexp.ts:
```diff
@@ -0,0 +1,466 @@
+  get isNil(): boolean {
+    return Module._TYPEOF(this.ptr) === SexpType.NILSXP;
+  }
+  getDollar(prop: string): RSexp | RawTypes {
+    throw new Error('The $ operator is not yet supported for this R object');
+  }
+  getIdx(idx: number): RSexp | RawTypes {
+    throw new Error('This R object cannot be indexed');
+  }
```
Should this be an overloaded `get()` function for `number` and `string` types?

lionel- commented on src/webR/webr-main.ts:
```diff
@@ -37,19 +24,30 @@ export class WebR {
 
-  // Backend delegation
-  async runRCode(_code: string) { return 'TODO'; };
+  evalRCode(code: string, envProxy?: RProxy): RProxy {
```
As argued elsewhere, maybe we don't need laziness?

lionel- commented on src/webR/webr-worker.ts:
```diff
@@ -69,49 +33,213 @@ const defaultOptions = {
+    let res = path.reduce(getProp, root) as RawTypes | RSexp | Function;
+    const parent = path.slice(0, -1).reduce(getProp, root) as RawTypes | RSexp;
+
+    // If requested, call the resulting object with the provided arguments
```
Ah since these methods are used to communicate across the channel, I guess the goal is to avoid a costly back and forth communication. Perhaps `getRObj()` should get a name that reflects this role? Such as `RObjRequest()`?

lionel- reviewed (commented):
Great progress! I've organised my main concerns into sections.


#### PR organisation

I think the implementation and review processes could have been more efficient by:

- Hoisting all the cosmetic changes to separate commits or PRs.

- Moving clearly separate changes such as the channel patch to separate PRs.

- First implementing one simple type instead of implementing many types at once. This gives an opportunity to discuss the API before tackling the other types. A draft could be an opportunity to discuss the API.

- Implementing the proxy stuff in a different PR. Probably a good candidate for some prior discussion as well.

  In the meantime, `evalRCode()` could have returned raw JS data and we could have added a new entry point `evalRCodeProxy()` later on.

I think we missed a discussion about the scope of the first release of webR. Perhaps the proxy stuff and the more complex data types could have been added in a future release, after we learned more about the constraints and needs of web applications that need to communicate with R.

Another thing that I feel rather uneasy about is that we are adding all this functionality when the project doesn't have unit tests set up yet. We have gotten by without tests until now, but for this sort of rich public API, each class should be tested extensively with high coverage.


#### Type unstability

Some methods or functions sometimes return completely different types depending on context/inputs. In general this sort of type unstability is error prone. It's fine to return different types when implementing polymorphism, but then the output type is directly related to the input type. Here I've seen cases where the output type varies depending on some internal state in the input, e.g. whether it has names or not.

So I think we should strive to return single types instead of type unions (typescript makes it easy to spot this). When multiple return types make sense, then it's better to provide different functions for each one.

- Methods returning `Rawtypes | RSexp`.
- Proxy methods returning `Rawtypes | RProxy`


#### Do we need to proxy?

While using Comlink and Synclink, I've fallen out of love with proxy objects because they often feel too magical, hard to use, and behave erratically in edge cases. With Synclink in particular, the abstraction completely broke in several cases which caused issues that were very hard to debug (it had to do with different serialisation/message passing semantics were used depending on state or input structure).

Here is what I found confusing while figuring out the semantics of these proxy objects:

- Need to know that `foo.bar.baz` returns a proxy. Also this always returns a proxy even when `bar` or `bar.baz` isn't defined, which I find confusing.

- Need to know that `await` must be used to _initiate_ a transfer. E.g. `await foo.bar.baz`. In my opinion it'd be more intuitive if a transfer was initiated via function call that you then await.

- A transfer or function call invokation may either return a proxy or JS data. I think we need a strong convention so that it's very clear when this will be the case, to avoid the pitfalls of type unstability.

- No type information on Proxy objects. It'd be nice to somehow have `RSexp` objects on the main side too, that programmers could use to program with typescript features enabled.

In addition, the current implementation of proxies returns `undefined` when an undefined property is invoked with `apply()`, but I think this isn't a downside of the proxy model and could be fixed.

So instead of proxying I wonder if it wouldn't be simpler and more predictable for users to have on the main thread side a generic `RObj` type with specific subtypes for typed R instead of keeping these on the worker side and accessing them through a proxy.

- `evalRCode()` would be an async function that returns a value implementing the generic `RObj` class. The caller could pass the result to specific class constructors like `RDouble`, `REnv`, or `RCharacter` to get access to more methods. The constructor would fail if the underlying object is incompatible with the class.

- There would be no implicit coercion to RAW objects. Instead, the user would have to explicitly call a typed method like `toJs()` or `toArray()`.

- The class methods like `env.ls()` would be async as well and communicate with the worker through the channel.

- Collection classes would have a `pluck()` method, e.g. `RVector.pluck()` and `REnv.pluck()`: Takes strings, symbols, numbers to reach into the data structure, always returns a promise, which always returns another `RObj` or `undefined` if the property doesn't exist.

- `RFunction.exec()` would be a method for the closure subclass to invoke it. I think it's better to use a special method to perform invokation. The function call syntax would hide that much is happening during invokation. Also it'll be more clearly async, and harder to miss that you need `await`.

`exec()` and `pluck()` are inspired by the corresponding purrr functions.

This setup would require more glue code between main and worker (could we somehow automate it?). However I think it would provide a much more predictable API for users and enable typescript linting and checking.


#### Other thoughts on API

- Avoid using getter functions for complex operations. The function call syntax is a helpful indication that non trivial computations or side effects may occur. Also, predicates like `isNil()` should have function call syntax I think, they look strange to me otherwise.

- Is `RCodeObj` useful? I see promises only useful when multiple promises are combined together. The caller could just hold the code as string and evaluate it when the result is needed. Perhaps we could simplify things by removing this type, if there is no clear advantage?

  If we can remove `RCodeObj`, and if we move away from type unstability, it seems like we could simplify `RTargetObj` to the `SEXPPTR` case?


#### Memory management

I was expecting to see some infrastructure or mechanism to deal with memory management and keep R objects alive while they are used on the JS side.

This concerns R objects wrapped by JS classes, but also intermediary objects returned by functions like `Rf_mkString()` which need to protected before calling back into the R API again.

georgestagg commented:
Thanks for your detailed thoughts and review. I've long form replied to your comments below, will add
some shorter ones in the inline comments, and have created some new PRs for you to take a look at when
you have the time.

> I think the implementation and review processes could have been more efficient by:
> Hoisting all the cosmetic changes to separate commits or PRs.
> Moving clearly separate changes such as the channel patch to separate PRs.

I've reworked all these changes into separate PRs to make review of each moving part easier. These are:
#44,#45,#46,#47,#48,#49, and #50. When we are finished I will close this PR unmerged.

> I think we missed a discussion about the scope of the first release of webR. Perhaps the proxy stuff and the more complex data types could have been added in a future release, after we learned more about the constraints and needs of web applications that need to communicate with R.

A fair comment. I do think it's definitely worth having at least some kind of JS interface in
the first release, though. I wouldn't feel comfortable having a versioned release without some kind of
way to get at R objects from JS, similar to Pyodide. On reflection, I can see that opinion is not
obvious or universal, but we are here now...

In practice, working with each type of object implemented so far felt natural as they interact with one another so deeply in R.

> Another thing that I feel rather uneasy about is that we are adding all this functionality
> when the project doesn't have unit tests set up yet.

I know, once compatibility with Node is merged back into main (#51) we can do this fairly well.
I have a plan to test with GitHub CI, Makefiles and comparing reference output.

> Type instability [...] In general this sort of type unstability is error prone [...] I think we
> should strive to return single types instead of type unions (typescript makes it easy to spot this).

OK, understood. I have reworked the `RObj` classes and have tried to be more stable with returning types.
Pyodide does this kind of JS object conversion implicitly, but I have removed the implicit conversion to JS objects in the interests of type stability, and arrays with just a single element are returned as
typed arrays.

I think at some point I would like to think further about under what circumstances (if any) should a
scalar atomic vector in particular return a number, rather than a typed array. Perhaps an optional
argument to turn that behaviour on?

I think it is useful to have a "magic" `toJs()` function, type unstable but returns a reasonable JS
conversion for simple types. Now, one asks for e.g. `toArray()`, `toObject()`, or `toJs()`
to convert/transfer, and `awaits` the resulting promise. `toJs()` is just an opinionated choice between `toArray` and `toObject` for each R object type.

There are some newly reworked `get` methods, corresponding to R's `[`, `[[` and `$` extracting operators,
and now actually calling these operators internally. These now always return another RObj or NULL.
It is useful to take in either a number or a string for some of these, allowing for name lookups.
I don't think overloading is as easy as creating two methods with the same name in JS/TS, so I have
handled this by taking in both types and testing between them with an if statement.

> While using Comlink and Synclink, I've fallen out of love with proxy objects because they
> often feel too magical, hard to use, and behave erratically in edge cases. [...]
> So instead of proxying I wonder if it wouldn't be simpler and more predictable for
> users to have on the main thread side a generic RObj type.

I have thought about this as part of the revision, and for the moment I'd like to stick with Proxies,
if that's OK. My main worry is the amount of new glue code that would be required. Perhaps it could be automated, but if it were easy I suppose that Comlink would already be doing it.

I think I would rather there be fewer types of request types over the channel and use proxies to
prepare messages. If we are stricter on typing. As the channels are async we'll always need lots of
`await`ing, so in the end things might look fairly similar to using a class on the main thread in any
case.

In the latest PRs the proxies are still there, but they work slightly differently. `evalRCode` is no longer lazy, and so `CODE` type target objects are no longer passed around. We still need to be able to pass around either `RAW` or `PTR` object types, because we might want to return either an `RObj` or `RawType` depending on the requested property or method, so there is still a little type instability over
the channel, but I've setup something in typescript to help with typing the proxy objects.

> Avoid using getter functions for complex operations. Also, predicates like isNil() should have function
> call syntax I think, they look strange to me otherwise.

This has been done in the latest PRs. One of the reasons I like the way the proxies work is that it
allows for easy chaining of properties like `obj.foo.bar.baz`. Functions instead must be awaited in a chain, `await (await (await obj.foo()).bar()).baz()`. Even so, I have reduced the use of getter
functions in `RObj`.

> Here is what I found confusing while figuring out the semantics of these proxy objects:
> Need to know that foo.bar.baz returns a proxy. Also this always returns a proxy even
> when bar or bar.baz isn't defined, which I find confusing.

Yes, that is a shame. I thought about this but could not see a way to remove that without losing some
functionality. At least, in the latest version, `await foo.bar` returns undefined if `bar` does
not exist, and `await foo.bar.baz` throws an error, even though `foo.bar.baz` without the await returns
a proxy.

> In my opinion it'd be more intuitive if a transfer was initiated via function call that you then await.
> A transfer or function call invokation may either return a proxy or JS data. I think we need a strong
> convention so that it's very clear when this will be the case, to avoid the pitfalls of type unstability.

I have tried to make it so that many methods return an `RObj` now, in particular for the `get` methods, which should help with this. I now picture statements like `await obj.foo[2].toJs()`, where everything before `toJs()` returns proxies, `toJs()` returns a promise, and the `await` waits for the transfer.
Since objects are no longer implicitly converted, there should be fewer instances of statements like `await obj.foo`, although there are still some that are possible where getters returning raw types are still around (e.g. `length`). Perhaps these can be removed all together in a further revision.

I hope that using the proxies will feel more intuitive now, in that to grab data you use functions like
`toJs()` and `toObject()`, which return promises to be awaited.

> No type information on Proxy objects. It'd be nice to somehow have RSexp objects on the main side too,
> that programmers could use to program with typescript features enabled.

I have tried to make this work with Typescript at the top of the file `proxy.ts`. The `RProxy<>` type is a mapped type that takes an `RObj` and constructs a type with the same object keys, but where returned
raw types are wrapped in a Promise, `RObjs` are wrapped in `RProxy`, and functions are changed so that the return type is wrapped in the same way.

I think this `RProxy<>` type constructs types that are compatible with what the proxies do. For
subclasses, users can access the extra methods by importing from `webR/robj` and `webR/proxy`, then setting the type with `as`.

For example, the following code on the main thread show various ways of interacting with a proxy, where typescript is aware of the need for the extra `await`s required to grab certain data.

```
  const r = await webR.evalRCode('c(1+2i, 3+4i)');
  await r.protect();
  const elem = await r.get(1);
  const elemJs = await elem.toJs();
  const lenPromise = (r as RProxy<RObjComplex>).length;
  const len = await lenPromise;
  const elem2 = r[2];
  const elem2Js = await r[2].toJs();
  const target = r._target;
```

Typescript can infer that `.protect()` returns a promise, `elem` is an R proxy object, `elemJs` is one of
the types in `RawType`, `len` is a number, `elem2` is a RProxy, `elem2Js` is a raw type, and target is
an `RTargetObj`. (At least, it seems to work in my editor...)

> In addition, the current implementation of proxies returns undefined when an undefined property
> is invoked with apply()

I think this is now fixed in the new PRs, an error is thrown.

> Collection classes would have a pluck() method, [...] RFunction.exec()

I have added these in addition to the proxy semantics described above on the `RObj` class. For now users
can choose which they prefer, and we can tighten things up in the future if we decide we dislike .get()
or proxy hooks.

The implmenation of `pluck()` gets objects using `get2` (as if using "[["), catching subscript out of
bounds and returning undefined in that case.

```
> r = await webR.evalRCode('list(a=1,b=list(c=2,d=c(5,6)))')
> val = await r.pluck('b','d',2)
> await val.toJs()
< Float64Array [6, buffer: ...]
```

`exec()` converts arguments into `RObjs` then invokes `_call`.

```
> c = await webR.evalRCode('cos');
> val = await c.exec([0.5, 0.75])
> await val.toJs()
< Float64Array(2) [0.8775825618903728, 0.7316888688738209, buffer: ...]
```

> Is RCodeObj useful?

`RCodeObj` is now gone, with the removal of lazy `evalRcode` evaluation.

> I was expecting to see some infrastructure or mechanism to deal with memory management and keep R
> objects alive while they are used on the JS side.

I've added `RObj.protect()` and `RObj.unprotect()` methods to protect and unprotect objects, and have tried
to be more strict with R object memory management in the `RObj` code. Users can now
`await rproxy.protect()` to protect an object. I had slightly misunderstood the GC and where protection was required, and hopefully things are looking a little better now.

georgestagg commented on src/webR/proxy.ts:
```diff
@@ -0,0 +1,83 @@
+          });
+        return r.then.bind(r);
+      }
+      return createRProxy(chan, target, [...path, prop]);
```
I've tried to use this early-exit style in my revisions, but do comment on the new PRs where I've slipped into old habits.

georgestagg commented on src/webR/webr-worker.ts:
```diff
@@ -69,49 +33,213 @@ const defaultOptions = {
+  const err = allocate(1, 'i32', 0);
+
+  let envObj = RSexp.wrap(RSexp.R_GlobalEnv);
+  if (env && env.type === RTargetType.CODE) {
+    envObj = wrapRTargetObj(env);
+  } else if (env && env.type === RTargetType.SEXPPTR) {
+    envObj = RSexp.wrap(env.obj);
+  } else if (env) {
+    throw new Error('Attempted to eval R code with an invalid raw env argument');
+  }
```
I've reworked the error handling in the new PRs. The issue was that when an error is thrown in the worker thread, if it is not caught it breaks the custom blocking loop and future execution is broken, requiring a page refresh.

In the new PR's I still try-catch errors, but in fewer (hopefully less confusing) places. I send them back to the main thread with a `write()` and handle them there. That way the custom blocking loop remains working.

georgestagg commented on src/webR/sexp.ts:
```diff
@@ -0,0 +1,466 @@
+}
+
+type logical = boolean | 'NA';
+class RSexpLogical extends RSexpVector {
```
This was a silly error. In the new commits I think atomic vectors are handled correctly.

lionel- commented:
> I've added RObj.protect() and RObj.unprotect() methods to protect and unprotect objects, and have tried to be more strict with R object memory management in the RObj code. Users can now await rproxy.protect() to protect an object. I had slightly misunderstood the GC and where protection was required, and hopefully things are looking a little better now.

Could we do automatic protection? I think we're going to see lots of randomly crashed instances of webR in the wild if we require users to protect manually. Manual protection would be fine for the internals of webR but even there automatic protection would keep us from introducing bugs.

lionel- commented:
It looks like this would be useful: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry

georgestagg commented:
Wow, great find! I was definitely worried about the lack of destructors in JS/TS. I will look into protecting on `new RObj` and unprotecting with the FinalizationRegistry.

I think you're probably right in general -- better to protect and then potentially leak memory rather than crashing.

georgestagg commented:
With the latest alternative `RProxy` implementation merged, this PR and branch is defunct.

The question of automatically protecting/preserving and unprotecting/releasing R Objects is still open, and will be tracked in Issue #52.
